File: pc/src/PhoneShell.Core/Services/TerminalHistoryBuffer.cs
namespace PhoneShell.Core.Services;

public sealed class TerminalHistoryBuffer
{
    private sealed record Chunk(long Seq, string Data);

    private readonly LinkedList<Chunk> _chunks = new();
    private readonly object _lock = new();
    private readonly long _maxChars;
    private long _nextSeq = 1;
    private long _totalChars;
    private const int DefaultPageChars = 20_000;

    public TerminalHistoryBuffer(int maxChars = 5_000_000)
    {
        _maxChars = Math.Max(0, maxChars);
    }

    public void Append(string data)
    {
        if (string.IsNullOrEmpty(data)) return;

        lock (_lock)
        {
            _chunks.AddLast(new Chunk(_nextSeq++, data));
            _totalChars += data.Length;
            TrimIfNeeded();
        }
    }

    public TerminalHistoryPage GetPage(long beforeSeq, int maxChars)
    {
        lock (_lock)
        {
            if (_chunks.Count == 0)
                return TerminalHistoryPage.Empty;

            var pageLimit = maxChars <= 0 ? DefaultPageChars : maxChars;
            if (_maxChars > 0)
                pageLimit = (int)Math.Min(pageLimit, _maxChars);
            LinkedListNode<Chunk>? node = null;

            if (beforeSeq <= 0)
            {
                node = _chunks.Last;
            }
            else
            {
                for (var cur = _chunks.Last; cur is not null; cur = cur.Previous)
                {
                    if (cur.Value.Seq < beforeSeq)
                    {
                        node = cur;
                        break;
                    }
                }
            }

            if (node is null)
                return TerminalHistoryPage.Empty;

            var parts = new List<string>();
            var used = 0;
            var oldestSeq = node.Value.Seq;

            for (var cur = node; cur is not null; cur = cur.Previous)
            {
                var data = cur.Value.Data;
                if (string.IsNullOrEmpty(data))
                {
                    oldestSeq = cur.Value.Seq;
                    continue;
                }

                if (used + data.Length > pageLimit)
                {
                    if (used == 0)
                    {
                        // Return the tail of a large chunk.
                        parts.Add(data[^pageLimit..]);
                        oldestSeq = cur.Value.Seq;
                        used = pageLimit;
                    }
                    break;
                }

                parts.Add(data);
                used += data.Length;
                oldestSeq = cur.Value.Seq;

                if (used >= pageLimit)
                    break;
            }

            if (parts.Count == 0)
                return TerminalHistoryPage.Empty;

            parts.Reverse();
            var payload = string.Concat(parts);
            var hasMore = _chunks.First!.Value.Seq < oldestSeq;
            var nextBeforeSeq = hasMore ? oldestSeq : 0;
            return new TerminalHistoryPage(payload, nextBeforeSeq, hasMore);
        }
    }

    private void TrimIfNeeded()
    {
        if (_maxChars <= 0)
            return;

        while (_totalChars > _maxChars && _chunks.Count > 0)
        {
            var first = _chunks.First!;
            _chunks.RemoveFirst();
            _totalChars -= first.Value.Data.Length;
        }
    }
}

public readonly record struct TerminalHistoryPage(string Data, long NextBeforeSeq, bool HasMore)
{
    public static TerminalHistoryPage Empty => new(string.Empty, 0, false);
}

File: linux2/src/terminal/terminal-history-buffer.ts
export interface TerminalHistoryPage {
  data: string;
  nextBeforeSeq: number;
  hasMore: boolean;
}

interface Chunk {
  seq: number;
  data: string;
}

const DEFAULT_PAGE_CHARS = 20000;

export class TerminalHistoryBuffer {
  private readonly chunks: Chunk[] = [];
  private totalChars = 0;
  private nextSeq = 1;
  private readonly maxChars: number;

  constructor(maxChars = 5000000) {
    this.maxChars = Math.max(0, maxChars);
  }

  append(data: string): void {
    if (!data) return;
    this.chunks.push({ seq: this.nextSeq++, data });
    this.totalChars += data.length;
    this.trimIfNeeded();
  }

  getPage(beforeSeq: number, maxChars: number): TerminalHistoryPage {
    if (this.chunks.length === 0) {
      return { data: '', nextBeforeSeq: 0, hasMore: false };
    }

    const limit = maxChars > 0 ? maxChars : DEFAULT_PAGE_CHARS;
    let idx = -1;

    if (beforeSeq <= 0) {
      idx = this.chunks.length - 1;
    } else {
      for (let i = this.chunks.length - 1; i >= 0; i--) {
        if (this.chunks[i].seq < beforeSeq) {
          idx = i;
          break;
        }
      }
    }

    if (idx < 0) {
      return { data: '', nextBeforeSeq: 0, hasMore: false };
    }

    const parts: string[] = [];
    let used = 0;
    let oldestSeq = this.chunks[idx].seq;

    for (let i = idx; i >= 0; i--) {
      const chunk = this.chunks[i];
      const data = chunk.data;
      if (!data) {
        oldestSeq = chunk.seq;
        continue;
      }

      if (used + data.length > limit) {
        if (used === 0) {
          parts.push(data.slice(data.length - limit));
          oldestSeq = chunk.seq;
          used = limit;
        }
        break;
      }

      parts.push(data);
      used += data.length;
      oldestSeq = chunk.seq;

      if (used >= limit) break;
    }

    if (parts.length === 0) {
      return { data: '', nextBeforeSeq: 0, hasMore: false };
    }

    parts.reverse();
    const payload = parts.join('');
    const hasMore = this.chunks[0].seq < oldestSeq;
    const nextBeforeSeq = hasMore ? oldestSeq : 0;
    return { data: payload, nextBeforeSeq, hasMore };
  }

  clear(): void {
    this.chunks.length = 0;
    this.totalChars = 0;
    this.nextSeq = 1;
  }

  private trimIfNeeded(): void {
    if (this.maxChars <= 0) return;
    while (this.totalChars > this.maxChars && this.chunks.length > 0) {
      const first = this.chunks.shift();
      if (first) this.totalChars -= first.data.length;
    }
  }
}
